src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * Quicksort implementation using Hoare partitioning
+ *
+ * <p>Run with:
+ *
+ * <p>$ ./gradlew run -Palgorithm=sorting.QuickSort3
+ *
+ * @author Atharva Thorve, aaathorve@gmail.com
+ */
+package com.williamfiset.algorithms.sorting;
+
+import java.util.Random;
+
+public class QuickSort3 implements InplaceSort {
+  private static Random random = new Random();
+
+  @Override
+  public void sort(int[] values) {
+    QuickSort3.quickSort3(values);
+  }
+
+  public static void quickSort3(int[] ar) {
+    if (ar == null) return;
+    QuickSort3.randomizedQuickSort(ar, 0, ar.length - 1);
+  }
+
+  // partiton array in such a way that all the elements whose value is equal to
+  // pivot are grouped together
+  private static int[] partition3(int[] a, int l, int r) {
+    int j, k;
+    if (r - l <= 1) {
+      if (a[r] < a[l]) {
+        int temp = a[l];
+        a[l] = a[r];
+        a[r] = temp;
+      }
+      j = l;
+      k = r;
+      int m1 = j;
+      int m2 = k;
+      int[] m = {m1, m2};
```

Comment:
Do you need m1 and m2 as temps here? I _think_ j and k would be copied by value, same below

Reply:
Yess okay I'll add that change.